File: src/CPP/src/leetcode/DivideTwoIntegers.cpp
/*
 * [Source] https://leetcode.com/problems/divide-two-integers/
 * [Difficulty]: Medium
 * [Tag]: Math
 * [Tag]: Binary Search
 */

#include <iostream>

using namespace std;

// [Solution]: Double the divisor by shift left until it is larger than dividend. Then minus and shift.
// [Corner Case]: divisor shift left leads to overflow, divisor is 0 or negative. abs fail for MIN_INT
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/mathandprob/DivideTwoIntegers.java
 */

/* Java solution
public class Solution {
    public int divide(int dividend, int divisor) { // dividend / divisor
        if(divisor == 0) {
            return Integer.MAX_VALUE; 
        }
        int result = 0; 
        if(dividend == Integer.MIN_VALUE) {
            result = 1; 
            dividend += Math.abs(divisor); // First minus one to avoid leak. 
        }
        if(divisor == Integer.MIN_VALUE) {
            return result; 
        }
        boolean negative = ((dividend ^ divisor) >>> 31 == 1) ? true : false; 
        dividend = Math.abs(dividend); 
        divisor = Math.abs(divisor); 
        int bitPos = 0; 
        while((dividend >> 1) >= divisor) {
            ++bitPos; 
            divisor <<= 1; // If dividend is not leak, divisor wont 
        }
        while(bitPos >= 0) { 
            if(dividend >= divisor) {
                dividend -= divisor; 
                result += 1 << bitPos; 
            }
            if(dividend == 0) {
                return negative ? -result : result; 
            }
            divisor >>= 1; 
            --bitPos; 
        }
        return negative ? -result : result; 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}

File: wwwroot/js/divide-two-integers.js
/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
    const int_max = 2147483648;
    let factor = 1;
    if ((dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0)) {
        factor = -1;
    }

    dividend = Math.abs(dividend);
    divisor = Math.abs(divisor);

    if (divisor > dividend) {
        return 0;
    }

    let cnt = 1;
    let val = divisor;
    while (val + val < dividend) {
        val = val + val;
        cnt = cnt + cnt;
    }

    let res = factor * (cnt + divide(dividend - val, divisor));
    if (res >= int_max) {
        return int_max - 1;
    } else {
        return res;
    }
};

document.getElementById("divide-two-integers").innerHTML =
    //divide(10, 3);
    //divide(4, 2);
    //divide(4, 4);
    //divide(1, 3);
    //divide(7, -3);
    divide(-2147483648, -1);
